tensorpipe/misc/dump_state_machine.cc

```diff
@@ -0,0 +1,126 @@
+#include <iostream>
+#include <regex>
+#include <unordered_set>
+
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+#include <clang/ASTMatchers/ASTMatchers.h>
+#include <clang/Frontend/FrontendActions.h>
+#include <clang/Tooling/CommonOptionsParser.h>
+#include <clang/Tooling/Tooling.h>
+#include <llvm/Support/CommandLine.h>
+
+using namespace clang::ast_matchers;
+using namespace clang::tooling;
+using namespace llvm;
+
+std::string exprToString(const clang::Expr& e) {
+  std::string statement;
+  raw_string_ostream stream(statement);
+  e.printPretty(stream, nullptr, clang::PrintingPolicy(clang::LangOptions()));
+  stream.flush();
+
+  return statement;
+}
+
+std::string cleanUp(const std::string& s) {
+  std::string res = s;
+  res = std::regex_replace(res, std::regex("(struct|class) [a-zA-Z_]+::"), "");
+  res = std::regex_replace(res, std::regex("this->"), "");
+  return res;
+}
+
+std::string escape(const std::string& s) {
+  std::string res = s;
+  res = std::regex_replace(res, std::regex("\\{"), "\\{");
+  res = std::regex_replace(res, std::regex("\\}"), "\\}");
+  res = std::regex_replace(res, std::regex(">"), "\\>");
+  res = std::regex_replace(res, std::regex("<"), "\\<");
+  res = std::regex_replace(res, std::regex("\\|"), "\\|");
+  return res;
+}
+
+class MethodPrinter : public MatchFinder::MatchCallback {
+  std::unordered_set<std::string> nodes_;
+
+  void addNode(const std::string& label) {
+    std::cout << label << " [label=<<b>" << label
+              << "</b>>,group=states,fontstyle=\"bold\"];" << std::endl;
+    nodes_.insert(label);
+  }
+
+ public:
+  virtual void run(const MatchFinder::MatchResult& Result) {
+    static int edgeCount = 0;
+
+    const clang::CallExpr& e = *Result.Nodes.getNodeAs<clang::CallExpr>("x");
+    std::string edgeId = "edge" + std::to_string(edgeCount++);
+    std::string fromId = cleanUp(exprToString(*e.getArg(1)));
+    std::string toId = cleanUp(exprToString(*e.getArg(2)));
+
+    if (nodes_.count(fromId) == 0) {
+      addNode(fromId);
+    }
+
+    if (nodes_.count(toId) == 0) {
+      addNode(toId);
+    }
+
+    std::string edgeColor = "orange3";
+    int edgeWeight = 100;
+    std::string cond = cleanUp(exprToString(*e.getArg(3)));
+    if (std::regex_search(cond, std::regex("^error_"))) {
+      edgeColor = "red3";
+      edgeWeight = 0;
+    }
+    if (std::regex_search(cond, std::regex("^!error_"))) {
+      edgeColor = "forestgreen";
+    }
+    cond = std::regex_replace(cond, std::regex(" \\&\\&"), "\\n");
```

Comment:
This means it won't handle more complicated expressions like `A || (B && C)` right? I think it's fine for now, we'll fix if if we'll ever need to.

Reply:
That's correct.